lib/galaxy/tools/__init__.py

```diff
@@ -2378,6 +2378,15 @@ def check_dataset_instance(input_dataset):
 
             map(check_dataset_instance, input_dataset_collection.dataset_instances)
 
+    def _add_datasets_to_history(self, history, elements):
+        datasets = []
+        for element_object in elements:
+            if getattr(element_object, "history_content_type", None) == "dataset":
+                datasets.append(element_object)
+
+        if datasets:
+            history.add_datasets(self.sa_session, datasets, set_hid=True)
```

Comment:
Should `set_hid` be a parameter for this method? Here is always `True`, but `False` was passed before in some cases (`ZipCollectionTool`, `MergeCollectionTool`, `FilterDatasetsTool`, `FlattenTool`, `SortTool`, `RelabelFromFileTool`, `TagFromFileTool`, `FilterFromFileTool`).
Not sure if that's connected with the failed API test: https://jenkins.galaxyproject.org/job/docker-api/12151/testReport/junit/api.test_tools/ToolsTestCase/test_map_over_collection_structured_like/

Reply:
It being False before was a bug I would say - we need to either copy the HDAs and assign them new HIDs or use the original HDAs. There have been some reports of duplicate HIDs #6350(?)  - and collection operations have probably been the problem.

I just clicked through on that issue and realized @mvdbeek knew this at least several days before me 😆.

Reply:
Oh this test failed because I rebased your changes and lost @mvdbeek's fix 😢.

Reply:
> It being False before was a bug I would say - we need to either copy the HDAs and assign them new HIDs or use the original HDAs.

It seems to me it should be possible to use the original HDAs for `ZipCollectionTool`, `MergeCollectionTool`, `FilterDatasetsTool`, `FlattenTool`, `SortTool`, `RelabelFromFileTool` and `FilterFromFileTool` classes.

Reply:
If you delete the new collection + contents - it deletes the old HDAs. You have a green collection with deleted HDAs and the user has no clue why. This was part of the original collection design that I really regret - the behavior makes sense from an API design perspective but it doesn't lead to a good UX. Collection contents shouldn't be HDAs - but if they are HDAs they should at least be in exactly one collection to make the UI work correctly. We've been on a path toward making collections more cohesive and I think it has been good.

We do not duplicate the datasets, just the HDAs - so we aren't duplicating space on disk for this or anything (expect BAI files - but the fix for that isn't reusing the HDAs I don't think). Just trying to improve the user experience.

If you want to open an issue to add an option to not duplicate the HDAs or something that is fine - we could discuss that more I think but certainly copying them and not giving them a new HID is a problem right?

Reply:
That makes sense, I've just tried to explore the "or use the original HDAs." part of your previous reply.
If you remove the line mentioned in https://github.com/galaxyproject/galaxy/pull/6370/files#r197497120 , I'll merge this.